### PR TITLE
fix make intangible hours editable

### PR DESCRIPTION
**Pull request by KaixiangGU**
**BE PR** https://github.com/OneCommunityGlobal/HGNRest/pull/267
This PR is about making intangible hours editable.
User -> View Profile -> Volunteering Times
BUG

✔️ How to test this PR:
1. Login with an admin account, manually edit intangible hours and save changes.
2. Submit a new intangible time entry to see if intangible hours keep adding.
3. Change intangible entry to tangible entry and vice versa to see if intangible hours will increase/decrease accurately.

### UPDATED on 1/28:
1. Prevent users input negative numbers in hours field. (Include Weekly Committed Hours)
2. Improve the UI on the volunteering times tab. (Labels are not vertically centred)
Before_UI_of_volunteering_times_tab
After_UI_of_volunteering_times_tab

### CODE REVIEW

**Review by rauleffting — CHANGES_REQUESTED** [review]
I reviewed your code. Think it worked quite well on changing intangible hours, both in the profile: 

ss2

And in the dashboard:

ss1

However, it's showing an error alert when saving changes. It seems not affecting the functionalities, but I think this is not supposed to appear:

ss3

I imagine this is a [backend](https://github.com/OneCommunityGlobal/HGNRest/pull/267) issue.

Once this small error is corrected, I believe we can approve the front and back PRs.

Thank you very much!

**Comment by KaixiangGU** [comment]
Hi @rauleffting, thanks for your review.
This error bug actually not related to my PR, it happened couple weeks ago, and it has already been fixed when I was working on intangible hours.
And my working branch was not the latest branch, that's why you saw this bug in my PR. To make sure this bug will not show up when my PR gets merged, I just pulled the latest changes from FE and BE, and everything works fine, hours can be edited without any errors.

**Comment by rauleffting** [comment]
@KaixiangGU Thanks to take a look. Now, the alert and error are no longer appearing. However, the 'intangible hours' field isn't getting updated anymore: 

https://user-images.githubusercontent.com/29555732/211036463-3a909a03-bcb0-4cca-bab4-a934376dcecd.mov

**Comment by rauleffting** [comment]
Hi @KaixiangGU. I tested here. The error is still occurring. After the first test, I also did a pull in front and back to test again and the error remains. Can you take a look? Thank you very much!

**Review by rauleffting — APPROVED** [review]
After pulled the development branch, it worked perfectly!

**Comment by rauleffting** [comment]
I've tested again and I found something. I followed your test list steps. Items 1 and 2 worked well, but Item 3 didn't. When changing tangible to intangible and vice versa at the 'Current Week Timelog', the value didn't update at the profile.

https://user-images.githubusercontent.com/29555732/212360724-ce6f649d-8569-4daf-99c8-3b098675cfea.mov

**Comment by KaixiangGU** [comment]
@rauleffting Hey, it should work now, please check it out. Thank you!

**Comment by rauleffting** [comment]
@KaixiangGU thank you very much! Your changes worked, as you can see below:

https://user-images.githubusercontent.com/29555732/212709740-da578398-ba46-4350-9559-edfd58aa1494.mov

however, as you saw at the end of the video, there's one other thing. When deleting the intangible/tangible time, the time doesn't change on the profile page. I'm not sure if it's related to your PR. So please correct me if I'm wrong.

Thank you very much!

**Comment by KaixiangGU** [comment]
Sure I can fix that! Thanks for the catch.

**Comment by KaixiangGU** [comment]
@rauleffting It should work now! Check it out!

**Review by rauleffting — DISMISSED** [review]
Your changes worked perfectly! I didn't see anything strange in your code as well. I'm going to approve it! Thank you very much!

**Review by RJBornToWin — CHANGES_REQUESTED** [review]
I see following issues while testing:
1. I am able to enter negative values using the down step button on the field. Upon save it doesn't get saved but shows success message instead

Screen Shot 2023-01-24 at 10 21 25 PM

2. On initial load, on development branch, my intangible hours was 1hr. But on your branch it shows empty box which is a major issue. Upon entering new value and saving stores only the new value. Old value just vanishes. This causes data integrity issues and will leave the application in data loss state.

On dev:
Screen Shot 2023-01-24 at 10 22 24 PM

But empty in your branch: Refer (1st) image.




NOTE: Upon save I still get error alert which i assume is a different problem irrespective of this PR.
Screen Shot 2023-01-24 at 10 26 09 PM

**Comment by KaixiangGU** [comment]
Hi @RJBornToWin Thanks for the review.

1. Sure, I will fix that negative number issue.

2. For old data, here is the problem. If you check current code base, I mean FE development branch. The way how to calculate intangible hours is that get all the time entries data, and sum up hours of all the intangible time entries. It actually didn't have any data stored in the database but it works in the past because intangible hours is not allowed to edit. 

But if it needs to be editable, I have to create an intangible hours property in the database and keep track of it when users make any changes. So, if this branch get merged, intangible hours will be start from 0.

Currently, I cannot think of any ways to keep old data and also make intangible hours editable. If you have any idea, please let me know! Thank you!

@one-community Hi Jae, do you really need keep all the old data of intangible hours? Or you okay with reset it from zero when it gets merged?

**Comment by one-community** [comment]
I'm ok if we lose the old intangible hours data.

**Comment by RJBornToWin** [comment]
@one-community : If you do not need the old intangible entry, how about removing the field ? We can add intangible entries from the user profile anyways.

**Comment by one-community** [comment]
I don't need the old intangible entry data because I have it all tracked on a spreadsheet, going back 10 years, for the entire management team. So I can manually enter it and then I want this to properly update it.

**Comment by KaixiangGU** [comment]
Hi @RJBornToWin , I made new changes as requested, I updated the description at the top of the page. Please check it out, thank you!

**Comment by RafaelCastilhoCastro** [comment]
Hi @KaixiangGU,
I tested the PR and both tangible and intangible fields are editable. The numbers are properly centered too.
I found some issues though:

1. Add intangible time -> access user profile -> change intangible hours to 0 -> Dashboard -> click checkbox tangible OR delete intangible time entry -> user profile -> Got negative intangible time.
I'm not sure if this is a bug since the fields will be editable, but I think its worth mentioning so it doesn't become a problem in the future.


2. The fields are accepting multiple zeros. It isn't causing problems with the data when saving according to my tests. I think it might be a good idea to fix it though.


3. The fields are accepting a large number of caracters before and after the decimal "."


Thanks.

**Comment by KaixiangGU** [comment]
Thank you @RafaelCastilhoCastro 
I'll take a closer look at #1 and #3.

For #2, I tried to solve this leading zeros issue in this PR. But cannot come up with a proper solution. If you have any idea about this, please let me know!

**Comment by tingedudu8** [comment]
Hi @KaixiangGU, I pulled both FE and BE and tested, but I found that I cannot make any change to Intangible time, is there something wrong with Intangible data? Thanks!

https://user-images.githubusercontent.com/70128933/216478195-2a111094-5ef2-4e12-aed5-71e7224468e5.mov

**Comment by KaixiangGU** [comment]
Hi @tingedudu8, It should work, Rafasel just tested it. From your video, it looks like BE is not running properly, I put BE link at the top of the page, please double check you're using the right PR. If there are any related errors on your console, please take a screenshot would be even better. Thank you!

**Comment by PrabhjotSembhi** [comment]
@KaixiangGU Hey, I am not seeing option to edit intangible hours
 
PR 635

**Comment by Filipe-Oliv** [comment]
HI @KaixiangGU 

1 - when I added the intangible hours in the volunteering times tab, an error popup appeared as shown in the image;

week 03 - 02-08-23 (6)
 
2 - when I returned to the main page, the box for the current week remained reset;

3 - when I added the main intangible time, the task was added to the task box and timalog, but the current week box kept with zero;

4 - when I returned to the volunteering times in the profile, there was an addition to the manual time added in the main with the time manually added in the volunteering times.

[week 03 - 02-08-23 (5).webm](https://user-images.githubusercontent.com/114119668/217651519-58bf07fe-f662-4e39-bce3-b927745dce1c.webm)

**Review by PrabhjotSembhi — COMMENTED** [review]
it's not working correctly when I check the box to switch intangible to tangible Time is not get updated. Please watch video 

https://user-images.githubusercontent.com/82662768/219467957-dd823194-98f2-45b6-becd-32536d9946bd.mp4

**Comment by Lunaliar** [comment]
I wasn't able to adjust the intangible time on other users' profiles as well as the current controlled profile. It seems like the form is not updating to the BE but Im not entirely sure. I really like the layout and formatting designs for this, however.
Here are some pics of my attempts to adjust the intangible time.
Screenshot from 2023-02-17 11-01-28
Screenshot from 2023-02-17 11-01-42


Seeing some functional components instead of class-based ones was really nice, I really like how you write your variables. I do have a few housekeeping things that may need a little attention.

In DeleteModal.jsx there are a few console.log functions on lines 23 & 28

In TimeEntry.jsx there is an unused variable on line 11 

In TimeEntryForm.jsx 
1) . There is an unused variable on line 24
2). There are 2 console.log functions on lines 397 and 444

In VolunteeringTimeTab.jsx there is a console.log function on line 198

**Comment by vitorsaa2k** [comment]
Im able to edit to intangible hours, but, there are a few problems:

1. I still get the error alert.
2. I can "accidentaly" make the intangible hours turn to a negative number.

https://user-images.githubusercontent.com/17365161/219722938-9771a1aa-30d6-4297-9953-30a195937059.mp4

As showed in the video:
1. I have created an intangible hours entry of 2hours
2. Edited the intangible hours on the profile to 0
3. When marking the intangible hours as tangible, the intangible hours on the profile turns into a negative number

**Comment by KaixiangGU** [comment]
@Lunaliar Thanks for the code check. I will delete those unused variables and console.logs.

**Comment by KaixiangGU** [comment]
Hi, @vitorsaa2k 
1. You don't have to worry about that error alert, that has been already fixed in the development branch.
2. I'm still working on the negative issue. I will push new commit when I fixed that.

**Review by suigunfjc — COMMENTED** [review]
Hi @KaixiangGU, I tested checking and unchecking the intangible timelog and the profile updated the intangible time correctly. Except for the error popup when saving the change, everything else looks pretty good to me.

https://user-images.githubusercontent.com/94319381/219909531-cf48cc95-a177-4b25-be46-efd4704cf935.mp4

**Comment by NyahJ** [comment]
Hi @KaixiangGU,

Follow your tesing suggestions:
First off, I got to manually edit intangible hours and save the change. Then, any new added intangible time entry was added. Last, it increased and decreased accurately for converting between tangible entry and intangible entry.
Since you mentioned the popup window is fixed, everything looks good from my side:) Thank you!
截屏2023-02-23 下午5 46 51

**Comment by harlleybastos** [comment]
@KaixiangGU I just tested here and it works but have one error, but once you already said on others comments that is already fixed on development, is okay !

https://user-images.githubusercontent.com/37200778/221199090-0d541220-ffb8-40d3-b454-59e050f38205.mp4

**Comment by dkamghem** [comment]
I am able to modify tangible time from 5 hours to 10 hours. 
Screen Shot 2023-03-02 at 9 34 22 PM (2)
Screen Shot 2023-03-02 at 9 34 43 PM (2)

**Comment by Lunaliar** [comment]
Hey, @KaixiangGU  I came back to this PR to debug my previous errors and was able to edit intangible hrs.
prior
post

**Comment by KaixiangGU** [comment]
Updated on 3/11
Prevent negative hours when edit time logs. Any negative hours will turn into zero.

**Comment by Filipe-Oliv** [comment]
Hi @KaixiangGU 

I managed to add and save negative hours and when I returned to the main my weekly hours goal also became negative.

File: src/components/Timelog/DeleteModal.jsx [hunk 1]
```diff
@@ -18,15 +18,29 @@ const DeleteModal = ({ timeEntry, userProfile }) => {
...
 
     dispatch(deleteTimeEntry(timeEntry));
+    //update hours
+    const formattedHours = parseFloat(timeEntry.hours) + parseFloat(timeEntry.minutes) / 60;
+    console.log(timeEntry);
+    if (!timeEntry.isTangible) {
+      userProfile.totalIntangibleHrs -= formattedHours;
+    } else {
+      const category = projectCategory ? projectCategory : taskClassification;
+      console.log(category);
```
Comment (rauleffting) [inline src/components/Timelog/DeleteModal.jsx:28]: Here too.

File: src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx [hunk 2]
```diff
@@ -197,272 +172,169 @@ const ViewTab = props => {
...
+    setUserProfile({
+      ...userProfile,
+      hoursByCategory: {
+        ...userProfile.hoursByCategory,
+        [key]: Number(value),
+      },
+    });
+    setChanged(true);
+  };
+  console.log(hoursByCategory);
```
Comment (rauleffting) [inline src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx:198]: Here too.

**Review by rauleffting — CHANGES_REQUESTED** [review]
Hi @KaixiangGU.

I've tested your PR and I was able to change the number inside the intangible time input. However, this is not affecting intangible time in other places, like the leaderboard. Is this supposed to behave like that?


https://user-images.githubusercontent.com/29555732/225363243-8ca0342f-e840-4166-9945-505390fda254.mp4


Still, you must have to delete some debug outputs. I've marked them for you.

Thank you very much!

**Comment by KaixiangGU** [comment]
@one-community Hi Jae, I think Raul just pointed out a good question. (Please check his video above)

If I manually edit the intangible hours, do I need also to change the number in effort bar under weekly timelog tabs? 

For the leaderboard, it only displays tangible hours, right? Editing intangible hours will not affect the leaderboard.

**Comment by one-community** [comment]
> If I manually edit the intangible hours, do I need also to change the number in effort bar under weekly timelog tabs?

No, that bar is only for the current week and should only show time logged via that page

> For the leaderboard, it only displays tangible hours, right? Editing intangible hours will not affect the leaderboard.

It shows only tangible as the bar but intangible are included in the number on the right... editing these numbers from the Profile page though should not affect that.

**Review by rauleffting — APPROVED** [review]
Hi, @KaixiangGU. I've tested your PR and it worked perfectly.

https://user-images.githubusercontent.com/29555732/228068576-0d2c380f-2584-4e0a-bf81-928dbf37f68d.mp4

I verified that you also made the changes that I requested.

Thank you very much! Approved!

**Comment by one-community** [comment]
Thanks everyone! Meeeeeeeerrrrrrgggggggiiiiinggggggggg! 👍

**Pull request merged**